### PR TITLE
S3 module

### PR DESCRIPTION
**Pull request by gurghet**
This is the new module. I’m pushing it because it works now but there are still quite a lot of caveats and `TODO`s.

Please disregard directory and file structure, I will fix them later.

The main things to review right now is the general architecture, the way state is advanced (this is the specialized, time based version) and especially how do we pass the `Transducer`.

Finally things that I could not solve are related to the default behavior: the key and the value in the simplest case are wrapped by case classes which are awkward, any ideas in that area is appreciated.

### CODE REVIEW

**Comment by gurghet** [comment]
Ok, there are still a few things off (no parallelism) but this could be an mvp.

**Comment by gurghet** [comment]
Never mind, I just realized it continuously publishes the latest state even if unchanged.

**Comment by gurghet** [comment]
@sirocchj actually I don't know if this is by design. Should the state be always republished when unchanged? I can't find the bit of code that would prevent this.

**Comment by sirocchj** [comment]
I need to actually look into this PR. You're correct saying it's "by design". But the design is also that _only_ the consumption of a new state triggers invocation of the fetch... so I'm not sure how this is supposed to work here

**Comment by gurghet** [comment]
One proposal could be that the state fold function becomes blocking. I doesn't do anything until the state changes. This will trow a wrench in the gears of the kafka loop.

**Review by barambani — APPROVED** [review]

File: s3/src/main/scala/tamer/s3/package.scala [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+  private final val logTask: Task[LogWriter[Task]] = log4sFromName.provide("tamer.s3")
+
+  private final val kafkaLayer: Layer[TamerError, Kafka] = Config.live >>> Kafka.live
+
+  type KeysR = Ref[List[String]]
+  val createRefToListOfKeys: UIO[KeysR] = Ref.make(List.empty[String])
+
+  private val defaultTransducer: Transducer[Nothing, Byte, Line] = ZTransducer.utf8Decode >>> ZTransducer.splitLines.map(Line)
+
+  def updateListOfKeys(
```
[the remark this answers is not in the document]
Comment (gurghet) [reply s3/src/main/scala/tamer/s3/package.scala:40]: this programme updates the list of keys so I think it's correct as is

**Comment by sirocchj** [comment]
Anything more we need to address here? Are we good to merge?

**Review by sirocchj — APPROVED** [review]
Let's get this in. It will give us visibility on the cross module concerns that we need to lift to core and we can start working on the ergonomics of the library for most of its use cases.

**Comment by gurghet** [comment]
Do you think we should have a more involved example app?

**Comment by sirocchj** [comment]
Probably worth it

**Comment by gurghet** [comment]
> Probably worth it

done in #235

**Pull request merged**